src/utilities.jl

```diff
@@ -52,13 +62,107 @@ end
 
 zstd_decompress(bytes::Vector{UInt8}) = transcode(ZstdDecompressor, bytes)
 
-####
-#### bytes/streams
-####
+#####
+##### read/write/bytes/streams
+#####
 
 jump(io::IO, n) = (read(io, n); nothing)
 jump(io::IOStream, n) = (skip(io, n); nothing)
 jump(io::IOBuffer, n) = ((io.seekable ? skip(io, n) : read(io, n)); nothing)
 
 unsafe_vec_uint8(x::AbstractVector{UInt8}) = convert(Vector{UInt8}, x)
 unsafe_vec_uint8(x::Base.ReinterpretArray{UInt8,1}) = unsafe_wrap(Vector{UInt8}, pointer(x), length(x))
+
+"""
+    read_byte_range(path, byte_offset, byte_count)
+
+Return the equivalent `read(path)[(byte_offset + 1):(byte_offset + byte_count)]`,
+but try to avoid reading unreturned intermediate bytes. Note that the
+effectiveness of this method depends on the type of `path`.
+"""
+function read_byte_range(path, byte_offset, byte_count)
+    return open(path, "r") do io
+        jump(io, byte_offset)
+        return read(io, byte_count)
+    end
+end
+
+read_byte_range(path, ::Missing, ::Missing) = read(path)
+
+write_full_path(path::AbstractString, bytes) = (mkpath(dirname(path)); write(path, bytes))
+write_full_path(path, bytes) = write(path, bytes)
+
+#####
+##### tables
+#####
+
+function table_has_supported_onda_format_version(table)
+    m = Arrow.getmetadata(table)
+    return m isa Dict && is_supported_onda_format_version(VersionNumber(get(m, "onda_format_version", v"0.0.0")))
+end
+
+# It would be better if Arrow.jl supported a generic API for nonstandard path-like types so that
+# we can avoid potential intermediate copies here, but its documentation is explicit that it only
```

Comment:
>  I think read_arrow_table(read(path)) does not usually make extra copies, right? it reads the whole file into memory as a `Vector{UInt8}, then Arrow.jl re-uses that byte buffer and basically exposes a lazy tabular view on top?

Yup. The part that's a bit more annoying IMO is the `write_arrow_table` workaround. I think it currently involves an unnecessary extra buffer...in theory it shouldn't need to, though. What I really want is a method that, given a `Table`, returns an `<:IO` object that directly references the table's memory and when read yields bytes in the caller's preferred Arrow file/IPC format (i.e. the `file=true` option for `write`). 

There might already be a good way to achieve that with the Arrow.jl API, and I just missed it lol

>  this API choice of Arrow.jl means we can't handle larger-then-memory tables or use mmaping, since we always need to read the whole table into memory.

Well, we still get `mmap`ing for local filesystem paths, but it would be really cool if e.g. AWSS3.S3Paths supported an `mmap`-like method that used byte-range reads under the hood 😎 

Generally, though, you'd split a larger-than-memory Arrow table into manageably chunked objects upon write, and then treat the objects as a one logical table/dataset via an API like https://arrow.apache.org/docs/python/dataset.html. In Julia land, I wonder how much we can get away with just using e.g. `Tables.partition` for things lol

Reply:
> The part that's a bit more annoying IMO is the write_arrow_table workaround. I think it currently involves an unnecessary extra buffer...in theory it shouldn't need to, though.

I see, yeah I've been using the same workaround and have also been annoyed by the need for it. I think maybe one fix is https://github.com/rofinn/FilePathsBase.jl/issues/113.

> I wonder how much we can get away with just using e.g. Tables.partition

That seems cool, I don't know how much as been figured out for using that kind of tool for out-of-memory data, as opposed to just lazily chaining tables that are already in-memory. But I haven't used it much yet.